src/Features/LanguageServer/Protocol/Handler/CodeActions/CodeActionResolveHandler.cs

```diff
@@ -26,86 +26,87 @@
+    {
+        var data = ((JToken)codeAction.Data!).ToObject<CodeActionResolveData>();
+        Assumes.Present(data);
+
+        // Fix All Code Action does not need further resolution since it already has the command callback
+        // when the action is initially created.
+        if (data.FixAllFlavors is not null)
+        {
+            return codeAction;
+        }
 
-        public async Task<LSP.CodeAction> HandleRequestAsync(LSP.CodeAction codeAction, RequestContext context, CancellationToken cancellationToken)
+        // We don't need to resolve a top level code action that has nested actions - it requires further action
+        // on the client to pick which of the nested actions to actually apply.
+        if (data.NestedCodeActions.HasValue && data.NestedCodeActions.Value.Length > 0)
         {
-            var data = ((JToken)codeAction.Data!).ToObject<CodeActionResolveData>();
-            Assumes.Present(data);
-
-            // Fix All Code Action does not need further resolution since it already has the command callback
-            // when the action is initially created.
-            if (data.FixAllFlavors is not null)
-            {
-                return codeAction;
-            }
-
-            // We don't need to resolve a top level code action that has nested actions - it requires further action
-            // on the client to pick which of the nested actions to actually apply.
-            if (data.NestedCodeActions.HasValue && data.NestedCodeActions.Value.Length > 0)
-            {
-                return codeAction;
-            }
-
-            var document = context.GetRequiredDocument();
-            var solution = document.Project.Solution;
-            var options = _globalOptions.GetCodeActionOptionsProvider();
-
-            var codeActions = await CodeActionHelpers.GetCodeActionsAsync(
-                document,
-                data.Range,
-                options,
-                _codeFixService,
-                _codeRefactoringService,
-                fixAllScope: null,
-                cancellationToken).ConfigureAwait(false);
-
-            Contract.ThrowIfNull(data.CodeActionPath);
-            var codeActionToResolve = CodeActionHelpers.GetCodeActionToResolve(data.CodeActionPath, codeActions, isFixAllAction: false);
-
-            // LSP currently has no way to report progress for code action computation.
-            var operations = await codeActionToResolve.GetOperationsAsync(
-                solution, CodeAnalysisProgress.None, cancellationToken).ConfigureAwait(false);
-
-            var edit = await CodeActionResolveHelper.GetCodeActionResolveEditsAsync(context, data, operations, cancellationToken).ConfigureAwait(false);
-
-            codeAction.Edit = edit;
             return codeAction;
         }
+
+        var document = context.TextDocument;
+        Contract.ThrowIfNull(document);
```

Comment:
can you add GetRequiredTextDocument() ?

Reply:
i would consider removing GetRequiredDocument as they are now all suspect (since presumably in normal scenarios they will be on a doc that is just an additional doc).

Reply:
GetRequiredDocument() is still needed by most Roslyn handlers as they can only operate on actual Documents.